### PR TITLE
AIRAVATA-3310: Implement uploaded file edit

### PR DESCRIPTION
**Pull request by dinukadesilva**
* Image preview and download (image/*)
* Text file edit and download (text/*). Text edit has no limit.
* Audio preview and download (audio/*)
* Video preview and download (video/*)
* Pdf preview and download (pdf)
* For any other files, the download is available

### CODE REVIEW

**Comment by dinukadesilva** [comment]
@machristie 
This is ready for review. 

Ref. test failures here, I see it's not due to my changes and looks like it has been before my changes.

File: django_airavata/apps/api/views.py [hunk 1]
```diff
@@ -1520,21 +1523,56 @@ def move_file(file_path, file_name, file_type):
             data_product = tus.move_tus_upload(uploadURL, move_file)
         return self._create_response(request, path, uploaded=data_product)
 
+    # Accept wither to replace file or to replace file content text.
+    def put(self, request, path="/", format=None):
+        # Replace the file if the request has a file upload.
+        if 'file' in request.FILES:
+            self.delete(request=request, path=path, format=format)
```
Comment (machristie) [inline django_airavata/apps/api/views.py:1530]: Can we have a data_products_helper.replace_data_product(request, data_product, file) instead of deleting and recreating? Reason is that we may want to re-register the DataProduct in the future and best to keep that logic together.

File: django_airavata/apps/api/data_products_helper.py [hunk 2]
```diff
@@ -129,18 +130,65 @@ def delete_dir(request, path):
     return datastore.delete_dir(request.user.username, path)
 
 
+def delete_user_file(request, path):
```
Comment (machristie) [inline django_airavata/apps/api/data_products_helper.py:133]: As much as possible, data_products_helper functions should work on DataProduct instances. There already is a `delete(request, data_product)` and there is an API to access it. See api/views.py, function delete_file (https://github.com/apache/airavata-django-portal/blob/develop/django_airavata/apps/api/views.py#L1009).

File: django_airavata/apps/api/views.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import json
 import logging
 import os
+import traceback
```
Comment (machristie) [inline django_airavata/apps/api/views.py:4]: Unused import

File: django_airavata/apps/api/views.py [hunk 4]
```diff
@@ -1520,21 +1523,56 @@ def move_file(file_path, file_name, file_type):
...
+    # Accept wither to replace file or to replace file content text.
+    def put(self, request, path="/", format=None):
+        # Replace the file if the request has a file upload.
+        if 'file' in request.FILES:
+            self.delete(request=request, path=path, format=format)
+            dir_path, file_name = split_dir_path_and_file_name(path=path)
+            self.post(request=request, path=dir_path, format=format, file_name=file_name)
+        # Replace only the file content if the request body has the `fileContentText`
+        elif request.data and "fileContentText" in request.data:
+            data_products_helper.update_file_content(request=request, path=path,
```
Comment (machristie) [inline django_airavata/apps/api/views.py:1535]: Please rename this function to update_data_product and make the second argument a DataProduct instance.

File: django_airavata/apps/api/views.py [hunk 5]
```diff
@@ -1520,21 +1523,56 @@ def move_file(file_path, file_name, file_type):
...
-        }
-        if uploaded is not None:
-            data['uploaded'] = uploaded
-        data['parts'] = self._split_path(path)
-        serializer = self.serializer_class(data, context={'request': request})
-        return Response(serializer.data)
+        if data_products_helper.dir_exists(request, path):
+            directories, files = data_products_helper.listdir(request, path)
+            data = {
+                'isDir': True,
```
Comment (machristie) [inline django_airavata/apps/api/views.py:1555]: Why is isDir needed? The intention of this view is that the path is always a directory.

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageTextEditViewer.vue [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        utils.FetchUtils.get(this.file.downloadURL, "", {
+          ignoreErrors: false,
+          showSpinner: true
+        }, "text").then((res) => {
+          this.fileContent = res;
+          this.setFileContentEditor(this.fileContent);
+        });
+      },
+      setFileContentEditor(value = "") {
+        this.editor = CodeMirror(document.getElementById("code"), {
```
Comment (machristie) [inline django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageTextEditViewer.vue:65]: Use a [ref](https://vuejs.org/v2/api/#ref) to get a reference to the code element, instead of doing getElementById, and remove the id from the div.

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageTextEditViewer.vue [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    },
+    data() {
+      return {
+        fileContent: "",
+        saved: true,
+        editor: null
+      };
+    },
+    mounted() {
+      this.setFileContent();
```
Comment (machristie) [inline django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageTextEditViewer.vue:47]: Make sure that the CodeMirror instance is torn down when the component is [destroyed](https://vuejs.org/v2/api/#destroyed). See for example how Vuejs can be used to wrap a jQuery component: https://vuejs.org/v2/examples/select2.html

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/entry-user-storage.js [hunk 8]
```diff
@@ -6,7 +6,7 @@ import VueRouter from "vue-router";
 
 const routes = [
   {
-    path: "*",
+    path: "/*",
```
Comment (machristie) [inline django_airavata/apps/workspace/static/django_airavata_workspace/js/entry-user-storage.js:9]: Can you explain why this change was made?

File: django_airavata/static/common/js/components/Uppy.vue [hunk 9]
```diff
@@ -30,6 +30,11 @@ export default {
       type: String,
       required: true
     },
+    xhrUploadMethod: {
```
Comment (machristie) [inline django_airavata/static/common/js/components/Uppy.vue:33]: Why was this added?

File: django_airavata/static/common/scss/main.scss [hunk 10]
```diff
@@ -117,6 +117,7 @@ body {
     flex-grow: 1;
     padding: 1rem;
     max-width: 1200px;
+    width: 100%;
```
Comment (machristie) [inline django_airavata/static/common/scss/main.scss:120]: Can you please explain why you added this? I don't see what this is doing.

File: django_airavata/apps/api/views.py [hunk 11]
```diff
@@ -1520,21 +1523,56 @@ def move_file(file_path, file_name, file_type):
...
+        if 'file' in request.FILES:
+            self.delete(request=request, path=path, format=format)
+            dir_path, file_name = split_dir_path_and_file_name(path=path)
+            self.post(request=request, path=dir_path, format=format, file_name=file_name)
+        # Replace only the file content if the request body has the `fileContentText`
+        elif request.data and "fileContentText" in request.data:
+            data_products_helper.update_file_content(request=request, path=path,
+                                                     fileContentText=request.data["fileContentText"])
+        else:
+            return Response(status=HTTP_403_FORBIDDEN)
```
Comment (machristie) [inline django_airavata/apps/api/views.py:1538]: I think a better error code here would be 400 since the request is malformed if it doesn't include a file upload or `fileContentText`.

File: django_airavata/apps/api/views.py (hunk 4 above)
Comment (dinukadesilva) [reply django_airavata/apps/api/views.py:1535]: I'm not too sure, I could get the DataProduct instance here.

File: django_airavata/apps/api/views.py (hunk 5 above)
Comment (dinukadesilva) [reply django_airavata/apps/api/views.py:1555]: I actually made this view to handle both files and directories.

File: django_airavata/static/common/js/components/Uppy.vue (hunk 9 above)
Comment (dinukadesilva) [reply django_airavata/static/common/js/components/Uppy.vue:33]: Actually this was added from the initial changes I've been doing to make the xhr method customizable. Though it was not needed later, I didn't remove that assuming it's helpful property to have

File: django_airavata/apps/api/data_products_helper.py (hunk 2 above)
Comment (dinukadesilva) [reply django_airavata/apps/api/data_products_helper.py:133]: Yes, I saw that too. But, I was trying to make this view generic for both files and directories. So, that I could make the request call from front end for both files and directories the same

File: django_airavata/static/common/scss/main.scss (hunk 10 above)
Comment (dinukadesilva) [reply django_airavata/static/common/scss/main.scss:120]: This was added to make `main-content` with 100% or 1200px max. Otherwise child element cannot be wrapped to the full width.

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/entry-user-storage.js (hunk 8 above)
Comment (dinukadesilva) [reply django_airavata/apps/workspace/static/django_airavata_workspace/js/entry-user-storage.js:9]: This can be removed

File: django_airavata/apps/api/views.py (hunk 1 above)
Comment (dinukadesilva) [reply django_airavata/apps/api/views.py:1529]: I didn't want to do so since I could use `delete` and `post` endpoints methods and take the advantage of its validations here.

**Comment by dinukadesilva** [comment]
@machristie 
I have updated the PR and added comments to which I need further clarifications.

File: django_airavata/apps/api/views.py (hunk 4 above)
Comment (machristie) [reply django_airavata/apps/api/views.py:1534]: You can get the data product URI using `data_products_helper.user_file_exists()` then load the data product with `request.getDataProduct()`.

However, I think this is fine, at least for now. The challenge is that the Airavata API uses data product URIs for user files, but we also need a path-based API in the Django portal for traversing and manipulating user directories. That's why there are two different kinds of REST endpoints, ones using data product URIs and ones using paths. Eventually we should update the Airavata API to handle directories as well as files then we can use Airavata API identifiers in the Django REST API in all places.

File: django_airavata/apps/api/views.py [hunk 12]
```diff
@@ -1535,7 +1534,7 @@ def put(self, request, path="/", format=None):
             data_products_helper.update_file_content(request=request, path=path,
                                                      fileContentText=request.data["fileContentText"])
         else:
-            return Response(status=HTTP_403_FORBIDDEN)
+            return Response(status=HTTP_404_NOT_FOUND)
```
Comment (machristie) [inline django_airavata/apps/api/views.py:1537]: I see you changed this to 404, but really it should be 400 since the request is missing required values.

File: django_airavata/apps/api/views.py (hunk 1 above)
Comment (machristie) [reply django_airavata/apps/api/views.py:1529]: Okay, yeah I think that's fine.

File: django_airavata/apps/api/views.py (hunk 5 above)
Comment (machristie) [reply django_airavata/apps/api/views.py:1554]: Okay, understood.

File: django_airavata/static/common/js/components/Uppy.vue (hunk 9 above)
Comment (machristie) [reply django_airavata/static/common/js/components/Uppy.vue:33]: If this isn't needed any longer, please remove.

File: django_airavata/static/common/scss/main.scss (hunk 10 above)
Comment (machristie) [reply django_airavata/static/common/scss/main.scss:120]: Specifically, can you describe where in the UI this makes a difference? For example, when a file is being edited in the codemirror editor?  I just want to verify if this is the best fix and also see if this change impacts any other views in the portal.

**Comment by machristie** [comment]
@dinukadesilva thanks for the updates. I've tried to add some clarification and I reviewed your changes.

I'm still getting an error when I try to create an experiment and select a file from storage. Can you please look at that too?

**Comment by dinukadesilva** [comment]
@machristie
I'm just curious about the error you are getting when creating an experiment. Due to may be some setting up issues, I actually cannot create any experiment in my localhost. Il try to get it fixed.

Till then,
1) Don't you get the same error without these changes
2) Can you post me the error log

**Comment by machristie** [comment]
Hi @dinukadesilva , you need to create an application that has an application input of type URI. Then you'll see the "Select file from storage OR upload" UI for that input when you create an experiment for that application.  Locally I tried Gaussian which has a single URI input.

When I select "Select file from storage" I get
```
Vue Global Error Handler TypeError: "this.userStoragePath is null"
    isDir webpack-internal:///./node_modules/cache-loader/dist/cjs.js?!./node_modules/babel-loader/lib/index.js!./node_modules/cache-loader/dist/cjs.js?!./node_modules/vue-loader/lib/index.js?!./static/django_airavata_workspace/js/components/storage/UserStoragePathViewer.vue?vue&type=script&lang=js&:125
    VueJS 4
    render webpack-internal:///./node_modules/cache-loader/dist/cjs.js?{"cacheDirectory":"node_modules/.cache/vue-loader","cacheIdentifier":"48bd918a-vue-loader-template"}!./node_modules/vue-loader/lib/loaders/templateLoader.js?!./node_modules/cache-loader/dist/cjs.js?!./node_modules/vue-loader/lib/index.js?!./static/django_airavata_workspace/js/components/storage/UserStoragePathViewer.vue?vue&type=template&id=9f8dcff2&:11
    VueJS 33
```

I also see these warnings:
```
[Vue warn]: Missing required prop: "username"

found in

---> <UserStoragePathViewer> at static/django_airavata_workspace/js/components/storage/UserStoragePathViewer.vue
       <UserStorageFileSelectionContainer> at static/django_airavata_workspace/js/components/storage/UserStorageFileSelectionContainer.vue
         <InputFileSelector> at static/django_airavata_workspace/js/components/experiment/input-editors/InputFileSelector.vue
           <FileInputEditor> at static/django_airavata_workspace/js/components/experiment/input-editors/FileInputEditor.vue
             <BFormGroup>
               <InputEditorFormGroup> at static/django_airavata_workspace/js/components/experiment/input-editors/InputEditorFormGroup.vue
                 <InputEditorContainer> at static/django_airavata_workspace/js/components/experiment/input-editors/InputEditorContainer.vue
                   <TransitionGroup>
                     <EditExperiment> at static/django_airavata_workspace/js/components/experiment/ExperimentEditor.vue
                       <CreateExperimentContainer> at static/django_airavata_workspace/js/containers/CreateExperimentContainer.vue
                         <MainLayout> at static/common/js/components/MainLayout.vue
                           <Root> vue.runtime.esm.js:620:15
```

```
[Vue warn]: Missing required prop: "storagePath"

found in

---> <UserStoragePathViewer> at static/django_airavata_workspace/js/components/storage/UserStoragePathViewer.vue
       <UserStorageFileSelectionContainer> at static/django_airavata_workspace/js/components/storage/UserStorageFileSelectionContainer.vue
         <InputFileSelector> at static/django_airavata_workspace/js/components/experiment/input-editors/InputFileSelector.vue
           <FileInputEditor> at static/django_airavata_workspace/js/components/experiment/input-editors/FileInputEditor.vue
             <BFormGroup>
               <InputEditorFormGroup> at static/django_airavata_workspace/js/components/experiment/input-editors/InputEditorFormGroup.vue
                 <InputEditorContainer> at static/django_airavata_workspace/js/components/experiment/input-editors/InputEditorContainer.vue
                   <TransitionGroup>
                     <EditExperiment> at static/django_airavata_workspace/js/components/experiment/ExperimentEditor.vue
                       <CreateExperimentContainer> at static/django_airavata_workspace/js/containers/CreateExperimentContainer.vue
                         <MainLayout> at static/common/js/components/MainLayout.vue
                           <Root> vue.runtime.esm.js:620:15
```

Do you have a testdrive.airavata.org account? I can point you to an application there that has a URI input so you can see how it is configured. You should then be able to create a similarly configured application in your local test environment.

**Comment by machristie** [comment]
Hi @dinukadesilva any update on this PR? Do you need any help?

**Comment by dinukadesilva** [comment]
@machristie 
Sorry for dragging this long.

Can you help me getting testdrive.airavata.org account and details. I didn't understand that clearly yet. Is it like a hosted airavata server which I could point my local django portal.

File: django_airavata/static/common/js/components/Uppy.vue (hunk 9 above)
Comment (dinukadesilva) [reply django_airavata/static/common/js/components/Uppy.vue]: Removed

File: django_airavata/static/common/scss/main.scss (hunk 10 above)
Comment (dinukadesilva) [reply django_airavata/static/common/scss/main.scss]: This is not needed. Removed

File: django_airavata/apps/api/views.py (hunk 12 above)
Comment (dinukadesilva) [reply django_airavata/apps/api/views.py]: Done

**Comment by machristie** [comment]
> Can you help me getting testdrive.airavata.org account and details. I didn't understand that clearly yet. Is it like a hosted airavata server which I could point my local django portal.

You can create an application in your local Django portal that has an input of type URI. Then you can create an experiment from that application. An input of type URI means it needs a URI that resolves to a file so when you create an experiment it will prompt you to select or upload a file. Then you should be able to replicate bug I mentioned above.

The testdrive.airavata.org account is just so you can see how an application (like Gaussian) is configured so you can create the same in your local Django instance for testing.  I'll see if I can get you read-only admin access.

**Comment by machristie** [comment]
@dinukadesilva you now have Read-Only Admin access in testdrive.airavata.org. You can look at the Gaussian application, specifically the interface, here: https://testdrive.airavata.org/admin/applications/Gaussian_9fe2b96c-9cb5-4319-8ccb-e680c4deaf78/interface

You don't need to configure an application exactly like this, this is just an example of an application that is configured with an input of type URI.

**Comment by dinukadesilva** [comment]
@machristie

I have made some changes. Can you have a look please.

I'm having trouble creating applications or experiments. Let me get that sorted first.

**Comment by machristie** [comment]
That changes you made look good @dinukadesilva 👍

**Comment by dinukadesilva** [comment]
@machristie
Has the issues being resolved now?

**Comment by machristie** [comment]
> @machristie
> Has the issues being resolved now?

Hi @dinukadesilva . I'm not quite sure what issues you are referring to, but I believe there is still an issue with creating experiments. You haven't fixed that yet, right?

**Comment by dinukadesilva** [comment]
@machristie
I fixed them. But not entirely sure since I have no way of testing. I have some issues recreating it and started a thread in the mailing list. Appreciate you help.

**Comment by machristie** [comment]
> @machristie
> I fixed them. But not entirely sure since I have no way of testing. I have some issues recreating it and started a thread in the mailing list. Appreciate you help.

Okay thanks @dinukadesilva , I'll test locally when I get and chance and will update you on how it goes.

**Comment by machristie** [comment]
@dinukadesilva I'm getting this error when I try to create an experiment:

```
Vue Global Error Handler TypeError: "this.$route is undefined"
    storagePath webpack-internal:///./node_modules/cache-loader/dist/cjs.js?!./node_modules/babel-loader/lib/index.js!./node_modules/cache-loader/dist/cjs.js?!./node_modules/vue-loader/lib/index.js?!./static/django_airavata_workspace/js/components/storage/UserStorageFileSelectionContainer.vue?vue&type=script&lang=js&:43
    VueJS 4
        get
        evaluate
        computedGetter
        get
    render webpack-internal:///./node_modules/cache-loader/dist/cjs.js?{"cacheDirectory":"node_modules/.cache/vue-loader","cacheIdentifier":"48bd918a-vue-loader-template"}!./node_modules/vue-loader/lib/loaders/templateLoader.js?!./node_modules/cache-loader/dist/cjs.js?!./node_modules/vue-loader/lib/index.js?!./static/django_airavata_workspace/js/components/storage/UserStorageFileSelectionContainer.vue?vue&type=template&id=58a24761&:15
    VueJS 7
 
Object { _uid: 38, _isVue: true, "$options": {…}, _renderProxy: Proxy, _self: {…}, "$parent": {…}, "$root": {…}, "$children": [], "$refs": {}, _watcher: {…}, … }
 render GlobalErrorHandler.js:43:15
```

The InputFileSelector.vue also uses the UserStorageFileSelectionContainer, but there is no `this.$route` there. I think maybe just a check to if `this.$route` is defined.

I wonder though why you need `storagePath` here and elsewhere? It might be simpler to create the `storagePath` from `userStoragePath.parts`. For example, you can essentially recreate storagePath with:
```javascript
["~"].concat(userStoragePath.parts).join("/") + "/"
```

**Comment by dinukadesilva** [comment]
@machristie 
Let me check the $route

I'm also very sure. But, even to get the `userStoragePath` I have to have the `storagePath` which is why I take that to many places as a prop.

**Comment by dinukadesilva** [comment]
@machristie I have done some fixes and need your review please.

**Comment by machristie** [comment]
Thanks @dinukadesilva I'll review it this week.

**Comment by machristie** [comment]
Hi @dinukadesilva . This is looking pretty good. I can select a file when creating an experiment now.

However, it looks like the entire storage UI is now embedded in the file selector. In the screenshot below, I've marked in red outline the part of the UI that shouldn't be displayed when selecting a file. Once that's fixed I can merge the PR.

Screenshot_2020-07-20 Screenshot

Thanks,

Marcus

**Comment by dinukadesilva** [comment]
@machristie I fixed it.

**Comment by machristie** [comment]
👍 Looks good to me! Thanks @dinukadesilva
